### DISCUSSION TITLE
Ctrl-r for shell history but also bulk select and remove from history

### DISCUSSION THREAD

**Discussion by rieje** [junegunn/fzf#3629] [Q&A] [answered]
Ctrl-r is super useful for shell history but along with Enter to insert the command to the prompt, I would like (within the same Ctrl-r interface, preferably) to also switch to interface where selecting an entry removes it from history. It should support selecting multiple entries and repeat filtering to select additional entries to remove (only except with e.g. Ctrl-c).

Any ideas on how to implement this if it's not implemented already?

**Comment by LangLangBart** [comment, marked as answer]
> [!CAUTION]
> Please note that this feature is experimental and should be used with caution. Some
> users have reported issues with this feature, so use at your own risk. Additionally, it is not
> being actively updated to match the latest changes in the original fzf version, which may result
> in missing features or functionality.


This setup requires 3 things:
- a modified `fzf-history-widget`
- the assigned `FZF_CTRL_R_OPTS` environment variable with a complex keybinding for deletion and
  reloading
- assigned `HISTFILE`, `HISTSIZE`, and `SAVEHIST` environment variables

> [!NOTE]
> The values assigned to `FZF_CTRL_R_OPTS` can also be incorporated into the modified
> `fzf-history-widget`. However, I found this method simpler to avoid complications with quotation
> marks.



If you place everything in your `.zshrc`, **after** sourcing `fzf` (`source <(fzf --zsh)`),
it should work seamlessly. The history will always be read from your `$HISTFILE`, thus I also set
 this option:

```bash
setopt SHARE_HISTORY  # Share history between all sessions.
```

More details on how the deletion works can be found in: #3522

```bash
modified-fzf-history-widget() {
  local selected
  setopt localoptions noglobsubst noposixbuiltins pipefail no_aliases no_bash_rematch 2> /dev/null
  # appends the current shell history buffer to the HISTFILE
  builtin fc -AI $HISTFILE
  # pushes entries from the $HISTFILE onto a stack and uses this history
  builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST
  selected="$(builtin fc -rl 1 |
    awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, "", cmd); if (!seen[cmd]++) print $0 }' |
    FZF_DEFAULT_OPTS="--height ${FZF_TMUX_HEIGHT:-40%} ${FZF_DEFAULT_OPTS-} -n2..,.. --scheme=history --bind=ctrl-r:toggle-sort,ctrl-z:ignore ${FZF_CTRL_R_OPTS-} --query=${(qqq)LBUFFER} --multi" $(__fzfcmd))"
  local ret=$?
	if [[ -n $selected ]]; then
    if [[ "$selected" =~ ^[[:blank:]]*[[:digit:]]+ ]]; then
	  builtin fc -pa "$HISTFILE"
	  zle vi-fetch-history -n "$MATCH"
    else # selected is a custom query, not from history
      LBUFFER="$selected"
    fi
  fi
  # Read the history from the history file into the history list
  builtin fc -R $HISTFILE
  zle reset-prompt
  return $ret
}
zle -N modified-fzf-history-widget
bindkey "^R" modified-fzf-history-widget


export FZF_CTRL_R_OPTS="$(
	cat <<'FZF_FTW'
--bind "ctrl-d:execute-silent(zsh -ic 'builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST; for i in {+1}; do ignore+=( \"${(b)history[$i]}\" );done;
	HISTORY_IGNORE=\"(${(j:|:)ignore})\";builtin fc -W $HISTFILE')+reload:builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST; builtin fc -rl 1 |
	awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, \"\", cmd); if (!seen[cmd]++) print $0 }'"
--bind 'enter:accept-or-print-query'
--header 'enter select · ^d remove'
--prompt ' Global History > '
FZF_FTW
)"


# 'ZDOTDIR' is a Parameter used by the shell, it refers to the location of your 
# shell startup files (see 'man zshparam')
export HISTFILE="${ZDOTDIR:-$HOME}"/.zsh_history
export HISTSIZE=12000
export SAVEHIST=10000
```

<img src="https://github.com/junegunn/fzf/assets/92653266/ee796237-ef15-4dd8-9edf-8686255b89b2" width="800">

EDIT: use `builtin` and read history when exiting `fzf` widget

**Reply by junegunn** [reply]
Thanks. Would it be possible to merge `execute-silent` and `reload`?

**Reply by LangLangBart** [reply]
The `fc -W` command needs to be run interactively (`zsh -ic`)

```bash
man zshbuiltins
# fc -ARWI [ filename ]
#        The fc command controls the interactive history mechanism.  Note
#        that reading and writing of history options is only performed if
#        the shell is interactive.  Usually this  is  detected  automati-
#        cally,  but  it can be forced by setting the interactive option
#        when starting the shell.
```

However, when I attempt to run it with `reload`, it blocks
```bash
: | fzf --bind "ctrl-b:reload:zsh -ic 'echo 4'"
```

**Reply by junegunn** [reply]
Huh, that's strange. I'll look into it.

---

This gives an even weirder result. 
```sh
: | fzf --bind "ctrl-b:reload:zsh -ic 'echo 4'; echo 5"
```

Seems related: https://github.com/jonas/tig/issues/951

(`execute-silent` doesn't set pgid, but `reload` does)

**Reply by andrevmatos** [reply]
Is this still working? When attempting this, after `reload` I get no result;
checking its stderr, I get `zsh:fc:1: no such event: 1`
I can't get any `fc` command to work inside reload.

**Reply by LangLangBart** [reply]
> I get `zsh:fc:1: no such event: 1`

Can you confirm if `HISTFILE`:
1.  is assigned?
2. exists?
3. has at least one entry?

```sh
cat $HISTFILE
```

> I can't get any fc command to work inside reload.

How is `SHELL` assigned ?

```sh
echo ${SHELL:-nothing}
```

If unassigned, `fzf` uses `sh -c` to execute commands, and this could point to `bash` on your operating system.
```sh
command ls -l /private/var/select/sh
# lrwxr-xr-x  1 root  wheel  9 Oct 20  2021 /private/var/select/sh -> /bin/bash
```

**Reply by andrevmatos** [reply]
I use NixOS with home-manager to set up zsh.
Yes, `$HISTFILE` is assigned, exists and has a tens of thousands of entries.
`$SHELL=/run/current-system/sw/bin/zsh`, as per NixOS default for zsh shell. I tried exporting it as `SHELL=zsh` as well, no change in behavior

**Reply by LangLangBart** [reply]
Tried to create a simple NixOS setup in a Dockerfile to test if a simplified `fzf` command with `fc` runs in there, and it seems to work. 

```sh
fzf --bind "start:reload:fc -p $HISTFILE;fc -rl 1"
```

---

Could you modify the `Dockerfile` to help reproduce the issue for me?

```docker
FROM nixos/nix:latest

# Install zsh and fzf
RUN nix-channel --update && \
    nix-env --install --attr nixpkgs.zsh nixpkgs.fzf

ENV SHELL /root/.nix-profile/bin/zsh
ENV HISTFILE /.zsh_history

RUN echo 'fzf --bind "start:reload:fc -p $HISTFILE;fc -rl 1"' >$HISTFILE

# Start zsh
CMD ["zsh", "-l", "-o", "INC_APPEND_HISTORY", "-o", "APPEND_HISTORY", "-o", "SHARE_HISTORY"]
```

```sh
docker build --tag nixos --file Dockerfile_nix .
docker run --interactive --tty --rm nixos:latest
```

**Reply by enoryw** [reply]
There's no way to manipulate the shell session history to be able to read and delete from it before it gets written to `HISTFILE` when the session ends, right? I'm interested in this fzf function but `INC_APPEND_HISTORY` and `SHARE_HISTORY` removes context of the history (e.g. with multiple terminal windows doing different things, since history gets added immediately, you no longer see commands grouped together by session in the history).

**Reply by mattmc3** [reply]
> There's no way to manipulate the shell session history to be able to read and delete from it before it gets written to `HISTFILE` when the session ends, right?

Per the manpages (`LESS="-p SHARE_HISTORY" man zshoptions`), `fc -RI` lets you sync history manually when you need it. So, if you don't want to set `SHARE_HISTORY`, you can easily add a test to `modified-fzf-history-widget` and sync history if SHARE_HISTORY isn't set with the following statement:

```
[[ -o SHARE_HISTORY ]] || fc -RI
```

> I'm interested in this fzf function but `INC_APPEND_HISTORY` and `SHARE_HISTORY` removes context of the history (e.g. with multiple terminal windows doing different things, since history gets added immediately, you no longer see commands grouped together by session in the history).

If you are worried about setting SHARE_HISTORY losing context, you can use `fc -lI` to show only commands in the current session, or `fc -lL` to show only the current session and the `$HISTFILE`, but not other sessions. `fc -l` will put an asterisk by commands from other sessions. Using `fc` properly can help alleviate many of the concerns with sharing history in Zsh. You can read more about `fc` here: `LESS="-p fc" man zshbuiltins`

**Reply by YasserKa** [reply]
Building on @LangLangBart solution; I don't recommend replacing the whole fzf default function, since it might get improved with upcoming features. For instance, his solution doesn't put into consideration the line wrapping feature that got added recently.

The recommended way is to add the lines using `awk` without dismissing any newly added code. Example:

```zsh
source <(fzf --zsh \
  | awk '
  /fzf-history-widget\(\)/ { 
  print; 
  flag = 1; 
  next 
}
/setopt/ { 
  print
if (flag) { 
  print "  # appends the current shell history buffer to the HISTFILE";
  print "  builtin fc -AI $HISTFILE";
  print "   # pushes entries from the $HISTFILE onto a stack and uses this history";
  print "  builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST";
} 
next
}
/zle reset-prompt/ { 
if (flag) { 
  print "  # Read the history from the history file into the history list"
  print "  builtin fc -R $HISTFILE"; 
  flag = 0 
} 
}
{ print }
'
)
```

**Reply by Drugoy** [reply]
@andrevmatos didn't reply to @LangLangBart about reproducing the issue, but I experience the same erroneous behavior while running on Debian 12 with regular zsh and latest fzf (0.57.0): upon hitting ctrl+d (while in ctrl-r mode) - the selected history item(s) get(s) deleted correctly, but the list doesn't get redrawn, so I have to hit enter or ctrl+c first and then ctrl-r again.
My `HISTFILE` env is set, cat'ing it shows the whole history.
I tried to manually execute the 'reload' function's shell code directly in the shell:
```shell
builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST; builtin fc -rl 1 | awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, "", cmd); if (!seen[cmd]++) print $0 }'
```
I don't fully understand what this command is supposed to do, but it returns thousands of backwards-numerated lines (up to the 1 at the bottom).
`zsh --version` says `zsh 5.9 (x86_64-debian-linux-gnu)`
`fzf --version` says `0.57.0 (0476a65)`

**Reply by Drugoy** [reply]
Same behavior on MacOS (with same versions of zsh and fzf).
I thought that maybe it has something to do with my .zshrc, but even without it, but with just `source <(fzf --zsh)` and the script above - the behavior is still the same, history reload doesn't happen.

**Reply by LangLangBart** [reply]
> but I experience the same erroneous behavior while running on Debian 12 with regular zsh and latest fzf (0.57.0): upon hitting ctrl+d (while in ctrl-r mode

Can you try this `docker` command, and after the terminal prompt appears, paste my original script? Then press Enter, type a dummy command like `ls`, press `Ctrl-R`, and then `Ctrl-D`. Does it reload? If it does try adding `--with-shell='/path/to/zsh/executable -fc'` to your `FZF_CTRL_R_OPTS` .

```bash
docker run --interactive --tty --rm alpine sh -c 'apk add fzf zsh;
{
  echo "source <(fzf --zsh)"
  echo "setopt interactivecomments share_history"
  echo "export FZF_DEFAULT_OPTS=\"--with-shell='\''/bin/zsh -fc'\''\""
} >> ~/.zshrc; zsh'
```

**Reply by Drugoy** [reply]
@LangLangBart  I tried, it works and reloads just fine in a container, but I either fail to follow your last instruction or it doesn't solve the problem outside the container.
How should the resulting `FZF_CTRL_R_OPTS` env look like, if zsh's binary path is `/usr/bin/zsh`?
like this?
```shell
export FZF_CTRL_R_OPTS="$(
	cat <<'FZF_FTW'
--with-shell='/usr/bin/zsh -fc'
--bind "ctrl-d:execute-silent(/usr/bin/zsh -ic 'builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST; for i in {+1}; do ignore+=( \"${(b)history[$i]}\" ); done;
	HISTORY_IGNORE=\"(${(j:|:)ignore})\";builtin fc -W $HISTFILE')+reload:builtin fc -p $HISTFILE $HISTSIZE $SAVEHIST; builtin fc -rl 1 |
	awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, \"\", cmd); if (!seen[cmd]++) print $0 }'"
--bind 'enter:accept-or-print-query'
--header 'enter select · ^d remove'
--prompt ' Global History > '
FZF_FTW
)"
```

**Reply by LangLangBart** [reply]
> --with-shell='/usr/bin/zsh -fc'

Yes, correct.

> or it doesn't solve the problem outside the container.

I will add a note that the feature is experimental and should be used at your own risk. If you like
to continue you could provide your zsh options. I will try to reproduce it on my system. Any other
special zsh oddities do you use, anything in `.zshenv`? A public .dotfiles repo would be helpful.

```bash
unsetopt KSH_OPTION_PRINT
setopt
```

 tested it with `debian:12` as well (works in the container)

```bash
docker run --interactive --tty --rm debian:12 sh -c 'apt update && apt install -y wget zsh;
  mkdir /fzf-bin
  wget https://github.com/junegunn/fzf/releases/download/v0.57.0/fzf-0.57.0-linux_amd64.tar.gz
  tar -xvf fzf-0.57.0-linux_amd64.tar.gz -C /fzf-bin
  {
    echo "export PATH=\"/fzf-bin:\$PATH\""
    echo "source <(fzf --zsh)"
    echo "setopt interactivecomments share_history"
    echo "export FZF_DEFAULT_OPTS=\"--with-shell='\''/bin/zsh -fc'\''\""
  } >> ~/.zshrc; zsh'
```

**Reply by Drugoy** [reply]
I've debugged my ~/.zshrc to the point where I cut almost everything and the issue was still reproducible with it (even in the container).
Then I noticed that I had
```shell
HISTFILE=~/.histfile
HISTSIZE=100000
SAVEHIST=100000
```
in it instead of
```shell
export HISTFILE=~/.histfile
export HISTSIZE=100000
export SAVEHIST=100000
```
and that's what was breaking the functionality.
I remember trying to narrow down the number of exported variables 'to reduce the pollution' (in favor of just setting variables when possible) and now it finally bit my ass...

Thanks you, @LangLangBart !

**Reply by LangLangBart** [reply]
> and that's what was breaking the functionality.

Thanks for pulling through (always good to know what the issue was). You might try to remove `--with-shell` from your `FZF_CTRL_R_OPTS`, it was suggested by me to ensure that the command is run with `zsh`. 

> [!NOTE]
> _fzf runs the command with `$SHELL -c` if `SHELL` is set, otherwise with `sh -c`_

**Reply by andrevmatos** [reply]
Thank you @LangLangBart , @Drugoy and @YasserKa .
That was it for me as well, and I managed to get it fixed by exporting the environment variables as well.
@YasserKa 's example seems to be more maintainable on the long run, but don't work with current fzf, which uses perl to handle multiline history entries. It'd be great to have a final/current version.
Thank you all.
